front/components/assistant/conversation/actions/AgentActionsPanel.tsx

```diff
@@ -1,18 +1,74 @@
-import { Spinner } from "@dust-tt/sparkle";
+import { ContentMessage, Markdown, Separator, Spinner } from "@dust-tt/sparkle";
 import React from "react";
 
 import { MCPActionDetails } from "@app/components/actions/mcp/details/MCPActionDetails";
 import { AgentActionsPanelHeader } from "@app/components/assistant/conversation/actions/AgentActionsPanelHeader";
 import { useConversationSidePanelContext } from "@app/components/assistant/conversation/ConversationSidePanelContext";
 import type { MCPActionType } from "@app/lib/actions/mcp";
 import { useConversationMessage } from "@app/lib/swr/conversations";
-import type { ConversationType, LightWorkspaceType } from "@app/types";
+import {
+  type ConversationType,
+  type LightWorkspaceType,
+  type MessageWithRankType,
+} from "@app/types";
+import {
+  isFunctionCallContent,
+  isReasoningContent,
+  isTextContent,
+} from "@app/types/assistant/agent_message_content";
 
 interface AgentActionsPanelProps {
   conversation: ConversationType | null;
   owner: LightWorkspaceType;
 }
 
+type Content =
+  | { kind: "reasoning"; content: string }
+  | { kind: "action"; action: MCPActionType };
+type Steps = Record<number, Array<Content>>;
+
+function groupContentsByStep(
+  fullAgentMessage: Extract<MessageWithRankType, { type: "agent_message" }>
+): Steps {
+  const steps: Steps = {};
+  for (const c of fullAgentMessage.contents) {
+    const step = c.step + 1;
+    if (!steps[step]) steps[step] = [];
+
+    if (isReasoningContent(c.content)) {
+      const reasoning = c.content.value.reasoning;
+      if (reasoning && reasoning.trim()) {
+        steps[step].push({ kind: "reasoning", content: reasoning });
+      }
+      continue;
+    }
+
+    if (isTextContent(c.content)) {
+      const thinkingRegex = /<thinking>([\s\S]*?)<\/thinking>/g;
```

Comment:
This won't support streaming right ? 

We have been doing this server side with 

https://github.com/dust-tt/dust/blob/bec488dc636437be8424178f7cd5a6b20e6d34e9/front/lib/api/assistant/agent_message_content_parser.ts#L28

There is more to it than just `<thinking>`